jcl/src/java.base/share/classes/openj9/internal/tools/attach/target/AttachHandler.java

```diff
@@ -522,6 +615,51 @@ static final class teardownHook extends Thread {
 			super("Attach API teardown"); //$NON-NLS-1$)
 		}
 
+		private void threadJoinHelper(Thread thread, long shutdownDeadlineNs) {
+			long timeout = 100;
+			int retryNumber = 0;
+			for (;;) {
+				long currentNanoTime = System.nanoTime();
+				if (currentNanoTime >= shutdownDeadlineNs) {
+					// already reached shutdownDeadlineNs
+					IPC.logMessage(thread + " already reached shutdownDeadlineNs : currentNanoTime = " + currentNanoTime ); //$NON-NLS-1$
+					break;
+				}
+				long tempTimeout = NANOSECONDS.toMillis(shutdownDeadlineNs - currentNanoTime);
+				if (tempTimeout < timeout) {
+					// not wait beyond shutdownDeadlineNs
+					timeout = tempTimeout;
+				}
+				IPC.logMessage(thread + ": currentNanoTime = " + currentNanoTime + ", tempTimeout = " + tempTimeout //$NON-NLS-1$ //$NON-NLS-2$
+						+ ", timeout = " + timeout); //$NON-NLS-1$
+				try {
+					thread.join(timeout);
+				} catch (InterruptedException e) {
+					IPC.logMessage(thread + ": join() interrupted"); //$NON-NLS-1$
+					break;
+				}
+				State state = thread.getState();
+				if (state == Thread.State.TERMINATED) {
+					// exit if thread is terminated
+					IPC.logMessage(thread + " is terminated, exit"); //$NON-NLS-1$
+					break;
+				}
+				IPC.logMessage(thread + ": state = " + state + ", timeout waiting for termination. Retry #" //$NON-NLS-1$ //$NON-NLS-2$
+						+ retryNumber + ", timeout = " + timeout); //$NON-NLS-1$
+				timeout *= 2;
+				retryNumber += 1;
+				if (thread == currentAttachThread) {
+					IPC.logMessage(thread + ": currentAttachThread requires AttachHandler.terminateWaitLoop() "); //$NON-NLS-1$
```

Comment:
Please remove the space at the end of the message here and on line 655.

Reply:
Removed